File: sem3/map/ToyLanguageInterpreter/src/utils/FileTableImpl.java
package utils;

import java.util.HashMap;
import java.util.Map;

/**
 * Created by mirko on 12/10/2016.
 */
public class FileTableImpl<K, V> implements FileTable<K, V> {
    private Map<K, V> fileTable;

    public FileTableImpl() {
        this.fileTable = new HashMap<K, V>();
    }

    @Override
    public void add(K key, V value) {
        fileTable.put(key, value);
    }

    @Override
    public V remove(K key) {
        return fileTable.remove(key);
    }

    @Override
    public boolean contains(K key) {
        return fileTable.containsKey(key);
    }

    @Override
    public V getValue(K key) {
        if (!contains(key)) throw new RuntimeException("error: key \"" + key + "\" doesn't exist in file table");
        return fileTable.get(key);
    }

    @Override
    public String toString() {
        StringBuilder string = new StringBuilder();
        string.append("FileTable = {");

        if (!fileTable.isEmpty()) string.append("\n");

        for (K key : fileTable.keySet()) {
            string.append("   " + key + " <- " + fileTable.get(key) + "\n");
        }

        string.append("}");
        return string.toString();
    }

    @Override
    public Iterable<Map.Entry<K, V>> getAll() {
        return fileTable.entrySet();
    }
}

File: sem3/map/ToySharpLanguageInterpreter/ToySharpLanguageInterpreter/src/utils/FileTableImpl.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace ToySharpLanguageInterpreter
{
	public class FileTableImpl<K, V> : FileTable<K, V>
	{
		private IDictionary<K, V> fileTable;

		public FileTableImpl()
		{
			this.fileTable = new Dictionary<K, V>();
		}

		public void add(K key, V value)
		{
			if (fileTable.ContainsKey(key)) fileTable[key] = value;
			else fileTable.Add(key, value);
		}


		public V remove(K key)
		{
			V value = fileTable[key];
			fileTable.Remove(key);
			return value;
		}

		public bool contains(K key)
		{
			return fileTable.ContainsKey(key);
		}

		public V getValue(K key)
		{
			if (!contains(key)) throw (new InterpreterException("error: key \"" + key + "\" doesn't exist in file table"));
			return fileTable[key];
		}

		public override string ToString()
		{
			string str = "";
			str += "FileTable = {";

			if (fileTable.Count() > 0) str += "\n";

			foreach (KeyValuePair<K, V> pair in fileTable.ToList())
			{
				str += "   " + pair.Key + " <- " + pair.Value + "\n";
			}

			str += "}";
			return str;
		}


		public List<KeyValuePair<K, V>> getAll()
		{
			return fileTable.ToList();
		}
	}
}